pytorch_lightning/strategies/bagua.py

```diff
@@ -0,0 +1,239 @@
+            )
+
+        super().__init__(
+            accelerator=accelerator,
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+            checkpoint_io=checkpoint_io,
+            precision_plugin=precision_plugin,
+        )
+
+        self._bagua_algorithm = algorithm
+        self._bagua_do_flatten = do_flatten
+        self._bagua_kwargs = bagua_kwargs
+
+    def setup_environment(self) -> None:
+        # start the other scripts
+        if not self.cluster_environment.creates_processes_externally:  # type: ignore[union-attr]
+            self._call_children_scripts()
+
+        self.setup_distributed()
+
+    def setup_distributed(self) -> None:
+        reset_seed()
+
+        # determine which process we are and world size
+        self.set_world_ranks()
+
+        self._init_bagua_distributed()
+
+    def _init_bagua_distributed(self) -> None:
+
+        self._set_node_environment_variables()
+        log.info(
+            "Initializing Bagua Distributed: "
+            f"GLOBAL_RANK: {self.global_rank}, "
+            f"MEMBER: {self.global_rank + 1}/{self.world_size}"
+        )
+
+        # need to set device first before initialize Bagua distributed environment
+        torch.cuda.set_device(self.local_rank)
+
+        if not is_initialized():
+            bagua.init_process_group()
+
+    def _set_node_environment_variables(self) -> None:
+        os.environ["MASTER_ADDR"] = self.cluster_environment.main_address  # type: ignore[union-attr]
+        os.environ["MASTER_PORT"] = str(self.cluster_environment.main_port)  # type: ignore[union-attr]
+        os.environ["RANK"] = str(self.global_rank)
+        os.environ["NODE_RANK"] = str(self.node_rank)
+        os.environ["WORLD_SIZE"] = str(self.world_size)
+        os.environ["LOCAL_RANK"] = str(self.local_rank)
+
+    def _check_qadam_optimizer(self) -> None:
+
+        has_qadam_optimizer = any([isinstance(opt, QAdamOptimizer) for opt in self.optimizers])
+
+        if not has_qadam_optimizer or len(self.optimizers) > 1 or len(self.lr_schedulers) > 1:
+            raise MisconfigurationException("Bagua QAdam can only accept one QAdamOptimizer and one LR Scheduler.")
+
+    def configure_ddp(self) -> None:
+        model = LightningBaguaModule(self.model)  # type: ignore[arg-type]
+        self._model = self._setup_model(model)
+
+        # start the background communication for async algorithm
+        if self.lightning_module.trainer.training and self._bagua_algorithm == "async":
+            self.model.bagua_algorithm.resume(self.model)  # type: ignore
+
+    def _setup_model(self, model: Module) -> BaguaDistributedDataParallel:
+        """Wraps the model into a Bagua distributed module."""
+
+        if self._bagua_algorithm == "qadam":
+            self._check_qadam_optimizer()
+            self._bagua_kwargs["q_adam_optimizer"] = self.optimizers[0]
```

Comment:
Can you move this block inside `_check_qadam_optimizer`? To group everything together